### PR TITLE
【Paddle Hackathon 5th No.93】add paddle pool3d op

### PR DESCRIPTION
**Pull request by AndPuQing**
### Details:
 - add paddle pool3d op convert support

### reference:

- https://www.paddlepaddle.org.cn/documentation/docs/en/2.4/api/paddle/nn/functional/max_pool3d_en.html
- https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/functional/avg_pool3d_en.html
- https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/functional/adaptive_avg_pool3d_en.html
- https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/functional/adaptive_max_pool3d_en.html

image

- https://github.com/openvinotoolkit/openvino/pull/20095

### CODE REVIEW

File: src/frontends/paddle/src/op/pool3d.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
+//*****************************************************************************
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+//*****************************************************************************
+
+#include "openvino/frontend/paddle/node_context.hpp"
+#include "openvino/opsets/opset6.hpp"
+#include "openvino/opsets/opset8.hpp"
```
Comment (meiyang-intel) [inline src/frontends/paddle/src/op/pool3d.cpp:9]: why include different version of opset header file. Can only include "default_opset.hpp" here?

**Comment by meiyang-intel** [comment]
Is there any document or code to describe the "pool3d". You just show the "max pool3d" in reference.

File: src/frontends/paddle/src/op/pool3d.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    PartialShape input_shape = data.get_partial_shape();
+
+    int32_t input_rank = static_cast<int32_t>(input_shape.rank().get_length());
+    PADDLE_OP_CHECK(node, input_rank >= 2, "input tensor rank must be greater than 2");
+
+    auto auto_pad = ov::op::PadType::EXPLICIT;
+    ov::Shape pad_begin, pad_end;
+    get_paddings(node, pad_begin, pad_end, auto_pad);
+
+    if (global_pooling || (adaptive && std::any_of(kernel_shape.begin(), kernel_shape.end(), [](int32_t i) {
```
Comment (meiyang-intel) [inline src/frontends/paddle/src/op/pool3d.cpp:95]: Is there test covering case of 'adaptive && one of kernel_shape == 1

File: src/frontends/paddle/src/op/pool3d.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                ov::opset6::Constant::create(ov::element::i64, {3}, {input_rank - 3, input_rank - 2, input_rank - 1});
+            return node.default_single_output_mapping({std::make_shared<ov::opset6::ReduceMax>(data, axes, true)},
+                                                      {"Out"});
+        } else {
+            auto axes =
+                ov::opset6::Constant::create(ov::element::i64, {3}, {input_rank - 3, input_rank - 2, input_rank - 1});
+            return node.default_single_output_mapping({std::make_shared<ov::opset6::ReduceMean>(data, axes, true)},
+                                                      {"Out"});
+        }
+    } else if (adaptive) {
```
Comment (meiyang-intel) [inline src/frontends/paddle/src/op/pool3d.cpp:109]: Is there test case covering this case?

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+
+def pool3d(name: str, x, attrs: dict):
+    import paddle
+
+    paddle.enable_static()
+
+    with paddle.static.program_guard(paddle.static.Program(), paddle.static.Program()):
+        node_x = paddle.static.data(name="x", shape=x.shape, dtype=data_type)
+        out = paddle.fluid.layers.pool3d(
```
Comment (OpenVINO-dev-contest) [inline src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py:21]: Since fluid api will be removed soon, shall we  use paddle.nn.functional.max_pool3d to validate this mapping ?@meiyang-intel

Comment (meiyang-intel) [reply src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py:21]: Good idea, @AndPuQing can you help to use the api in paddle.nn.functional?

File: src/frontends/paddle/src/op/pool3d.cpp (hunk 1 above)
Comment (AndPuQing) [reply src/frontends/paddle/src/op/pool3d.cpp]: Update. Most of them have used the default opset, But when I use default_opset MaxPool, strange errors will appear.

image

File: src/frontends/paddle/src/op/pool3d.cpp (hunk 2 above)
Comment (AndPuQing) [reply src/frontends/paddle/src/op/pool3d.cpp:95]: Already covered

File: src/frontends/paddle/src/op/pool3d.cpp (hunk 3 above)
Comment (AndPuQing) [reply src/frontends/paddle/src/op/pool3d.cpp:111]: Already covered

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py (hunk 4 above)
Comment (AndPuQing) [reply src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py]: Already fixed

**Comment by OpenVINO-dev-contest** [comment]
Pls rename your PR as  ```【Paddle Hackathon 5th No.93】add paddle pool3d op ```

**Comment by AndPuQing** [comment]
> Pls rename your PR as `【Paddle Hackathon 5th No.93】add paddle pool3d op `

Done

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            fetchlist=[out],
+            inputs=[x],
+            outputs=[outs[0]],
+            target_dir=sys.argv[1],
+        )
+
+    return outs[0]
+
+
+def adaptive_pool2d(name: str, x, attrs: dict):
```
Comment (meiyang-intel) [inline src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py:60]: should the name be "adaptive_pool3d"?

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+            # input=data_NDHWC, # shape: [2, 4, 4, 4, 3]
+            "pool_size": [3, 3, 3],
+            "pool_type": pooling_type,
+            "pool_stride": [3, 3, 3],
+            "pool_padding": [1, 2, 1],
+            "ceil_mode": False,
+            "exclusive": True,
+            "data_format": "NDHWC",
+        }
+        # NOT support data_format = "NDHWC" now
```
Comment (meiyang-intel) [inline src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py:183]: why not support the NDHWC?

File: src/frontends/paddle/src/op/pool3d.cpp (hunk 1 above)
Comment (meiyang-intel) [reply src/frontends/paddle/src/op/pool3d.cpp]: https://github.com/openvinotoolkit/openvino/blob/master/src/core/include/openvino/op/max_pool.hpp , MaxPool has two kinds of class v1 and v8. It's not accuracy to use opset6 for MaxPool here. When you use opset6, the MaxPool should use v1. The default_opset will use the latest class v8. The v8 MaxPool has two outputs, https://docs.openvino.ai/2022.3/openvino_docs_ops_pooling_MaxPool_8.html#doxid-openvino-docs-ops-pooling-max-pool-8 . Maybe it is the problem?

Comment (AndPuQing) [reply src/frontends/paddle/src/op/pool3d.cpp]: Thank you, I figure out, Now Fixed

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py (hunk 5 above)
Comment (AndPuQing) [reply src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py]: fixed

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py (hunk 6 above)
Comment (AndPuQing) [reply src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py:183]: The main reason is because maxpool does not directly provide `data_format` parameter.

It seems that two transpose ops need to be added to support

Comment (meiyang-intel) [reply src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py:183]: Seems you have the idea to support it, can you implement it?

Comment (AndPuQing) [reply src/frontends/paddle/tests/test_models/gen_scripts/generate_pool3d.py]: Done, Now we support "NDHWC" input. And I add `paddle_avg3dPool_test10` to test the `padding` when `NDHWC` input

- https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/functional/max_pool3d_en.html

File: src/frontends/paddle/src/op/pool3d.cpp [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        }
+    }
+
+    if (data_format == "NDHWC") {
+        pool_outputs[0] = std::make_shared<default_opset::Transpose>(
+            pool_outputs[0],
+            std::make_shared<default_opset::Constant>(ov::element::i64, Shape{5}, std::vector<int64_t>{0, 2, 3, 4, 1}));
+    }
+
+    if (pool_outputs.size() == 1) {
```
Comment (meiyang-intel) [inline src/frontends/paddle/src/op/pool3d.cpp:208]: The number of outputs should be the same as output number of paddle op, not the number of openvino pool op.

Comment (AndPuQing) [reply src/frontends/paddle/src/op/pool3d.cpp:208]: The tricky problem is that paddle maxpool op will have two outputs (Regardless of return_mask, [return_mask is implemented by python](https://github.com/PaddlePaddle/Paddle/blob/release/2.5/python/paddle/nn/functional/pooling.py#L1451))

But, when op is adapte && global_pool, openvino will use ReduceMax op. It cannot get the index. 

do you have any good advice?

Comment (meiyang-intel) [reply src/frontends/paddle/src/op/pool3d.cpp:208]: I check the python code, if want to return mask it will use 'max_pool3d_with_index'. So 'max_pool3d_with_index' has two outputs and 'pool3d' has one output. So in op_table.cpp, you can map 'max_pool3d_with_index' and 'pool3d' to the two different openvino frontend ops, they are different.

**Comment by ilya-lavrenov** [comment]
@meiyang-intel looks like the PR is green

**Comment by meiyang-intel** [comment]
> @meiyang-intel looks like the PR is green

This PR is not completed yet. Still have a problem.

File: src/frontends/paddle/src/op/pool3d.cpp (hunk 7 above)
Comment (AndPuQing) [reply src/frontends/paddle/src/op/pool3d.cpp]: I have updated it, use a new function to handle `max_pool_with_index`

File: src/frontends/paddle/src/op/pool3d.cpp [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+            pool_outputs[0],
+            std::make_shared<default_opset::Constant>(ov::element::i64, Shape{5}, std::vector<int64_t>{0, 2, 3, 4, 1}));
+    }
+
+    auto output_name = node.get_output_names();
+    if (output_name.size() == 1) {
+        return NamedOutputs{{"Out", {pool_outputs[0]}}};
+    } else {
+        PADDLE_OP_CHECK(node, pool_outputs.size() == 2, "returnmask can`t use with global_pooling");
+        return NamedOutputs{{"Out", {pool_outputs[0]}}, {"Mask", {pool_outputs[1]}}};
```
Comment (meiyang-intel) [inline src/frontends/paddle/src/op/pool3d.cpp:215]: Which case still need to output "Mask" execept 'max_pool3d_with_index'?

Comment (AndPuQing) [reply src/frontends/paddle/src/op/pool3d.cpp]: It seem could be removed, Any return_mask will to `max_pool3d_with_index`

**Comment by meiyang-intel** [comment]
@AndPuQing , CI test failed, pls fix it.

**Review by meiyang-intel — APPROVED** [review]

**Comment by AndPuQing** [comment]
@meiyang-intel any progress?
